### PR TITLE
Nullable ints

### PR DESCRIPTION
**Pull request by rvagg**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,9 +1,23 @@
 module github.com/whyrusleeping/cbor-gen
 
-go 1.12
+go 1.18
```
Comment (rvagg) [inline go.mod:3]: I can undo this if needed, I just wanted to use generics but it's a few lines of code difference if you'd prefer to keep this lower.

File: gen.go [hunk 2]
```diff
@@ -938,7 +946,38 @@ func emitCborUnmarshalStructField(w io.Writer, f Field) error {
...
+	b, err := cr.ReadByte()
+	if err != nil {
+		return err
+	}
+	if b != cbg.CborNull[0] {
+		if err := cr.UnreadByte(); err != nil {
+			return err
+		}
+		maj, extra, err = {{ ReadHeader "cr" }}
+		var extraI int64
```
Comment (Stebalien) [inline gen.go:959]: nit: move this down.

File: gen.go [hunk 3]
```diff
@@ -938,7 +946,38 @@ func emitCborUnmarshalStructField(w io.Writer, f Field) error {
...
+		case cbg.MajNegativeInt:
+			extraI = int64(extra)
+			if extraI < 0 {
+				return fmt.Errorf("int64 negative overflow")
+			}
+			extraI = -1 - extraI
+		default:
+			return fmt.Errorf("wrong type for int64 field: %d", maj)
+		}
+		{{ .Name }} = &extraI
```
Comment (Stebalien) [inline gen.go:978]: type alias support?

File: gen.go [hunk 4]
```diff
@@ -938,7 +946,38 @@ func emitCborUnmarshalStructField(w io.Writer, f Field) error {
...
+	if b != cbg.CborNull[0] {
+		if err := cr.UnreadByte(); err != nil {
+			return err
+		}
+		maj, extra, err = {{ ReadHeader "cr" }}
+		var extraI int64
+		if err != nil {
+			return err
+		}
+		switch maj {
```
Comment (Stebalien) [inline gen.go:963]: Can we avoid duplicating this logic? I.e., reuse the same code below and have a second "if pointer" check?

File: gen.go (hunk 3 above)
Comment (Stebalien) [reply gen.go:978]: I.e.: `&{{ .TypeName }}(extraI)` (or whatever compiles.

Comment (rvagg) [reply gen.go]: did this locally today but didn't push it

File: gen.go (hunk 4 above)
Comment (rvagg) [reply gen.go:963]: 👍

File: gen.go [hunk 5]
```diff
@@ -938,29 +946,38 @@ func emitCborUnmarshalStructField(w io.Writer, f Field) error {
...
+			}
+			extraI = -1 - extraI
+		default:
+			return fmt.Errorf("wrong type for int64 field: %d", maj)
+		}
 
-	{{ .Name }} = {{ .TypeName }}(extraI)
-}
+		{{ .Name }} = {{ if .Pointer }}(*{{ .TypeName }})(&extraI){{ else }}{{ .TypeName }}(extraI){{ end }}
+		{{ if .Pointer }}}{{ end }}}
```
Comment (Stebalien) [inline gen.go:980]: 🐍 

I'd put the `}`s on their own lines.

File: gen.go [hunk 6]
```diff
@@ -938,29 +946,38 @@ func emitCborUnmarshalStructField(w io.Writer, f Field) error {
...
+				return fmt.Errorf("int64 negative overflow")
+			}
+			extraI = -1 - extraI
+		default:
+			return fmt.Errorf("wrong type for int64 field: %d", maj)
+		}
 
-	{{ .Name }} = {{ .TypeName }}(extraI)
-}
+		{{ .Name }} = {{ if .Pointer }}(*{{ .TypeName }})(&extraI){{ else }}{{ .TypeName }}(extraI){{ end }}
```
Comment (Stebalien) [inline gen.go:979]: style nit: I'd break this over multiple lines (can duplicate the assignment, it's not much)

**Review by Stebalien — APPROVED** [review]

File: gen.go (hunk 5 above)
Comment (rvagg) [reply gen.go]: was trying to minimise whitespace change in output, latest commit preserves it though while spacing these out

**Comment by Stebalien** [comment]
Merge conflicts, but it looks ready otherwise.

**Comment by rvagg** [comment]
updated, ping @whyrusleeping

**Pull request merged**